### PR TITLE
First attempt to implement dependency chains for gfx-hal

### PR DESCRIPTION
**Pull request by zakarumych**
This change is [Reviewable](https://reviewable.io/reviews/omni-viral/chains/2)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# gfx-chain
+
+This crates provides means for the users of low-level graphics API of `gfx-hal` to reason about what passes uses what resources and how those resources are accessed.
```
Comment (kvark) [inline README.md:3]: "use"

File: README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+With this information `gfx-chain` may automatically derive synchronization commands required
+before and after each pass.
+
+## Overview
+
+### Resources
+
+Entry point to work with `gfx-chain` is `Resource` trait.
+User should implement if for resource types. It is expected to be wrappers for `gfx_hal::Backend::Buffer` and `gfx_hal::Backend::Image`.
```
Comment (kvark) [inline README.md:12]: "implement if"

Comment (kvark) [inline README.md:12]: "It is expected to be wrappers"

File: README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+### Passes
+
+Passes are logically coupled operations performed by GPU.
+For graphics they are identical to Vulkan's subpass.
+Pass uses resources the same way each time it is invoked.
+Each pass can declare a set of resource-categories it uses. Pass can provide information about how resources from each category is used, how they are accessed and in what layout resources are expected to be.
```
Comment (kvark) [inline README.md:19]: "resources from each category is used"

File: README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+Additionally pass specify on which command queue operations are executed.
+
+### Chain
+
+Access, layout, usage and queue info (links) collected from all passes associated with each resource-category form dependency chain. Chain automatically derive synchronization commands (pipeline barriers and semaphores) required for resource-category. During command recording chain can be used to insert required barriers for specified concrete resources from the category. For submission chain can be asked if semaphores must be signaled or waited upon.
```
Comment (kvark) [inline README.md:24]: "Chain automatically derive"

Comment (kvark) [inline README.md:24]: Need to spray over "a"/"the" to make this more readable

File: src/buffer/mod.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+impl Access for BufferAccess {
+    fn none() -> Self {
+        Self::empty()
+    }
+    fn all() -> Self {
+        Self::all()
+    }
+    fn is_write(&self) -> bool {
+        self.contains(Self::TRANSFER_WRITE) || self.contains(Self::SHADER_WRITE)
```
Comment (kvark) [inline src/buffer/mod.rs:15]: should probably be rewritten with a composite mask, e.g.
```rust
let mask = Self::TRANSFER_WRITE | Self::SHADER_WRITE |...; //TODO: upstream this
!(*self & mask).is_empty()
```

File: src/chain/chain.rs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+use hal::pso::PipelineStage;
+
+use chain::pass::PassLinks;
+use chain::link::{Acquire, Link, LinkSync, Release};
+use resource::{Access, Layout, Resource, Usage};
+
+/// Unique identifier for resource dependency chain.
+/// Multiple resource can be associated with single chain
+/// if all passes uses them the same way.
```
Comment (kvark) [inline src/chain/chain.rs:12]: "all passes uses them"

File: src/chain/chain.rs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        ChainId(index, PhantomData)
+    }
+
+    /// Get index value.
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Resource chain for set of resources.
```
Comment (kvark) [inline src/chain/chain.rs:29]: similarly, use more prepositions, e.g. "Resource chain for **a** set of resources"

File: src/chain/chain.rs [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+/// All resources in set are expected to be used only in ways specified by this chain.
+/// Each link of the chain corresponds to some render-pass or similar entity.
+#[derive(Clone, Debug)]
+pub struct Chain<A, L, U, S, W> {
+    usage: U,
+    links: Vec<Option<Link<A, L, S, W>>>,
+}
+
+impl<A, L, U, S, W> Chain<A, L, U, S, W>
```
Comment (kvark) [inline src/chain/chain.rs:38]: we could parametrize the implementation by `R` right away, since the only method it has (`build`) defines `R`, and we can derive `A/L/U` with their bounds from it

File: src/chain/chain.rs [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+    L: Layout,
+    U: Usage,
+{
+    /// Build chain from links defined by passes.
+    pub fn build<P, R, F>(id: ChainId<R>, mut usage: U, passes: P, mut new_semaphore: F) -> Self
+    where
+        P: IntoIterator,
+        P::Item: Borrow<PassLinks<R>>,
+        R: Resource<Access = A, Layout = L, Usage = U>,
+        F: FnMut() -> (S, W),
```
Comment (kvark) [inline src/chain/chain.rs:50]: need some docs on what `S` / `W` are

File: src/chain/chain.rs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+            let (link, after) = link_after.split_first_mut().unwrap();
+
+            // Skip non-existing
+            let link = if let Some(link) = link.as_mut() {
+                link
+            } else {
+                continue;
+            };
+
+            if let Some(next) = after
```
Comment (kvark) [inline src/chain/chain.rs:127]: let's rewrite it to break out earlier and not indent the whole branch body

File: src/chain/chain.rs [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+                break;
+            }
+        }
+
+        for index in 0..count {
+            let (before, link_after) = links.split_at_mut(index);
+            let (link, after) = link_after.split_first_mut().unwrap();
+
+            // Skip non-existing
+            let link = if let Some(link) = link.as_mut() {
```
Comment (kvark) [inline src/chain/chain.rs:121]: let's use `match`

File: src/chain/chain.rs [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+                usage = usage | link.usage;
+                Link {
+                    queue: pass.queue,
+                    access: link.access,
+                    layout: link.layout,
+                    stages: link.stages,
+                    merged_access: link.access,
+                    merged_layout: link.layout,
+                    merged_stages: link.stages,
+                    acquire: LinkSync::None(Acquire),
```
Comment (kvark) [inline src/chain/chain.rs:68]: Let's rename the enum variant to avoid confusion

File: src/chain/chain.rs [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+        R: Resource<Access = A, Layout = L, Usage = U>,
+        F: FnMut() -> (S, W),
+    {
+        let mut links: Vec<Option<Link<A, L, S, W>>> = Vec::new();
+
+        // Walk over passes
+        for pass in passes {
+            let pass = pass.borrow();
+            // Collect links from passes.
+            links.push(pass.links.iter().find(|link| link.id == id).map(|link| {
```
Comment (kvark) [inline src/chain/chain.rs:58]: should we be filtering here instead of finding? and if the `id` is unique, we could consider having `pass.links` to be `HashMap<ChainId, Link>`

File: src/chain/chain.rs [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+
+            // Skip non-existing
+            let link = if let Some(link) = link.as_mut() {
+                link
+            } else {
+                continue;
+            };
+
+            // Get next existing link
+            if let Some(next) = after
```
Comment (kvark) [inline src/chain/chain.rs:90]: please explain what is going on here. Unclear to me:
  - the code appears to be going through each link and attempts to merge it with every other link?
  - shouldn't we remove the links after merging?
  - shouldn't we only try to merge with links before the current one?
  - why going twice?

File: src/chain/chain.rs [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+                            layout: src_layout..dst_layout,
+                            stages: src_stages..dst_stages,
+                        };
+                    }
+                    _ if link.queue.family() == next.queue.family() || !next.access.is_read() => {
+                        // Same family.
+                        // Or different family but content is discarded.
+                        // Ownership transfer should be skipped. See note at
+                        // https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#synchronization-queue-transfers
+                        let (signal, wait) = new_semaphore();
```
Comment (kvark) [inline src/chain/chain.rs:168]: why do we generalize over semaphores?

File: src/chain/chain.rs [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+impl<A, L, U, S, W> Chain<A, L, U, S, W> {
+    /// Get reference to link by index.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if no link exists at this index.
+    ///
+    pub fn link(&self, index: usize) -> &Link<A, L, S, W> {
```
Comment (kvark) [inline src/chain/chain.rs:229]: nit: could also implement `ops::Index` and such, but `index` should probably be wrapped up into something more type safe

File: src/chain/chain.rs [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+    pub fn link_mut(&mut self, index: usize) -> &mut Link<A, L, S, W> {
+        self.links[index].as_mut().unwrap()
+    }
+
+    /// Get reference to first link before specified index.
+    pub fn prev(&self, index: usize) -> Option<&Link<A, L, S, W>> {
+        self.links[..index]
+            .iter()
+            .rev()
+            .filter_map(Option::as_ref)
```
Comment (kvark) [inline src/chain/chain.rs:248]: what is the purpose of storing `None` values in `self.links`?

File: src/chain/chains.rs [hunk 18]
```diff
@@ -0,0 +1,71 @@
+use std::ops::{Index, IndexMut};
```
Comment (kvark) [inline src/chain/chains.rs:1]: need to rename the module to `res_chains` to avoid confusion

File: src/chain/chains.rs [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+use super::chain::{Chain, ChainId};
+use super::pass::PassLinks;
+use resource::{Resource, Usage};
+
+/// Type alias for `Chain` bound to specific resource type.
+pub type ResourceChain<R: Resource, S, W> = Chain<R::Access, R::Layout, R::Usage, S, W>;
```
Comment (kvark) [inline src/chain/chains.rs:8]: I feel incoming Rust warnings about type generics not being enforced. That doesn't mean this code needs fixing, but rather - Rust needs to be.

File: src/chain/chains.rs [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    /// `new_semaphore` - function to allocate new semaphore as a pair of tokens. One for signalling. Another for waiting.
+    /// 
+    pub fn new<F>(
+        passes: &[PassLinks<R>],
+        mut new_semaphore: F,
+    ) -> ResourceChains<R, S, W>
+    where
+        F: FnMut() -> (S, W),
+    {
+        let total_count = passes.iter().flat_map(|pass| pass.links.iter().map(|link| link.id.index())).max().unwrap_or(0);
```
Comment (kvark) [inline src/chain/chains.rs:34]: let's split the equation to multiple lines

File: src/chain/chains.rs [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+        ResourceChains {
+            chains: (0..total_count)
+                .map(|i| Chain::build(ChainId::new(i), R::Usage::none(), passes, &mut new_semaphore))
+                .collect(),
+        }
+    }
+
+    /// Get reference to chain by id.
+    pub fn chain(&self, id: ChainId<R>) -> &ResourceChain<R, S, W> {
```
Comment (kvark) [inline src/chain/chains.rs:44]: why are these needed given that you already implement `Index`?

File: src/chain/link.rs [hunk 22]
```diff
@@ -0,0 +1,238 @@
+use std::ops::Range;
```
Comment (kvark) [inline src/chain/link.rs:1]: why are the main sources in `src/chain/xxx.rs` as opposed to just `src/xxx.rs`?

File: src/chain/link.rs [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+
+use resource::{Access, Layout, Resource};
+use queue::QueueId;
+
+#[derive(Clone, Copy, Debug)]
+pub(super) struct Acquire;
+#[derive(Clone, Copy, Debug)]
+pub(super) struct Release;
+
+pub(super) trait Semantics {
```
Comment (kvark) [inline src/chain/link.rs:16]: need some docs on what it does

File: src/chain/link.rs [hunk 24]
```diff
@@ -0,0 +1,238 @@
...
+    fn src_dst(this: QueueFamilyId, other: QueueFamilyId) -> (QueueFamilyId, QueueFamilyId);
+}
+
+impl Semantics for Acquire {
+    fn src_dst(this: QueueFamilyId, other: QueueFamilyId) -> (QueueFamilyId, QueueFamilyId) {
```
Comment (kvark) [inline src/chain/link.rs:21]: nit: could also use ranges

File: src/chain/link.rs [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+impl Semantics for Release {
+    fn src_dst(this: QueueFamilyId, other: QueueFamilyId) -> (QueueFamilyId, QueueFamilyId) {
+        (this, other)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(super) enum LinkSync<A, L, S, M> {
```
Comment (kvark) [inline src/chain/link.rs:33]: please describe what the generic arguments mean here

File: src/chain/link.rs [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+
+    /// Pipeline barrier.
+    Barrier {
+        access: Range<A>,
+        layout: Range<L>,
+        stages: Range<PipelineStage>,
+    },
+
+    /// Signal / Wait for semaphore.
+    Semaphore { semaphore: S },
```
Comment (kvark) [inline src/chain/link.rs:45]: it does appear that the enum could also be represented by a struct with optional fields:
```rust
struct LinkSys {
  barrier: Option<Barrier>,
  semaphore: Option<S>,
  transfer: Option<QueueId>,
}
```

File: src/chain/link.rs [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+        self.access
+    }
+
+    /// Get resource layout for the link.
+    pub fn layout(&self) -> L {
+        self.layout
+    }
+
+    /// Record acquire barrier if required.
+    pub fn acquire<R, C>(
```
Comment (kvark) [inline src/chain/link.rs:206]: similarly, the `acquire` and `release` could be moved out into a separate `impl` block parametrized by `R`

File: src/chain/mod.rs [hunk 28]
```diff
@@ -0,0 +1,9 @@
+mod chain;
```
Comment (kvark) [inline src/chain/mod.rs:1]: nonono, `chain/chain` and `chain/chains` needs revising!

**Review by kvark — COMMENTED** [review]
A ton of good stuff here! And sorry about the late review.
Hopefully, my notes are going to be useful.

File: README.md (hunk 2 above)
Comment (mjadczak) [reply README.md:12]: > The entry point for working with `gfx-chain` is the `Resource` trait. Users should implement it for resource types. It is expected to be a wrapper for ...

File: README.md (hunk 1 above)
Comment (mjadczak) [reply README.md:3]: > This crate provides the means for users of the low-level graphics API provided by `gfx-hal` ...

File: README.md (hunk 3 above)
Comment (mjadczak) [reply README.md:19]: > A Pass uses resources the same way each time it is invoked.
> A Pass can provide information about how resources from each category are used ...

File: README.md (hunk 4 above)
Comment (mjadczak) [reply README.md:24]: > form dependency chains. Chains automatically derive ... required for the resource-category (for the current resource category? for the appropriate resource category?)

> During command recording, a chain can be used to insert ...

> During submission, the chain can be asked ...

File: src/chain/chain.rs (hunk 6 above)
Comment (mjadczak) [reply src/chain/chain.rs:12]: > A unique identifier ...

> if all passes use them

File: src/chain/chain.rs (hunk 8 above)
Comment (zakarumych) [reply src/chain/chain.rs:38]: Compiller wouldprobably complain that `R` is not used in type signature.

File: src/chain/chain.rs (hunk 14 above)
Comment (zakarumych) [reply src/chain/chain.rs:90]: I'd like to remove links after merging them, but link is associated with single pass. I'll try to do that in later design iteration.

File: src/chain/link.rs (hunk 26 above)
Comment (zakarumych) [reply src/chain/link.rs:45]: Except that having some queue without barrier and semaphore is nonsence.
